[RFC] Allow Pkg.test to pass command-line arguments to test/runtests.jl

DilumAluthge opened the pull request:
Fixes #518 
Fixes #981 
Fixes #1241   

---

In a source build of Julia, you can run specific test sets as such:

```
./julia test/runtests.jl bitarray math
```

In a binary install, you can accomplish the same thing with:

```
Base.runtests(["bitarray", "math"])
```

`Base.runtests(tests)` accomplishes this (see [here](https://github.com/JuliaLang/julia/blob/b7774095f62e47bb2c8467b8a328075ab5a3fb89/base/util.jl#L813)) by passing the contents of the first positional argument `test` as the command line arguments to `test/runtests.jl`. 

However, as far as I can tell, there is currently no functionality built into `Pkg` that allows you to pass any arguments to a package's `test/runtests.jl` file when running the package tests.

This pull request adds two features:
- Pass flags to the `julia` process that runs a package's tests. The flags are specified in the `julia_flags` keyword argument to `Pkg.test`.
- Pass positional command-line arguments to `test/runtests.jl`. The arguments are specified in the `test_args` keyword argument to `Pkg.test` and are available inside `test/runtests.jl` as `ARGS[1]`, `ARGS[2]`, etc.

Example usage:

```julia
# equivalent to running `julia test/runtests.jl`
Pkg.test("HelloWorld")
```

```julia
# equivalent to running `julia test/runtests.jl foo bar baz`
Pkg.test("HelloWorld"; test_args = ["foo", "bar", "baz"])
```

```julia
# equivalent to running `julia --math-mode=ieee test/runtests.jl`
Pkg.test("HelloWorld"; julia_flags = ["--math-mode=ieee"])
```

```julia
# equivalent to running `julia --math-mode=ieee test/runtests.jl foo bar baz`
Pkg.test("HelloWorld"; test_args = ["foo", "bar", "baz"], julia_flags = ["--math-mode=ieee"])
```

---

This pull request only adds the functionality for passing the command line arguments to `test/runtests.jl`. It is up to the author of a package to add the logic for processing the command line arguments inside their `test/runtests.jl` file. However, this should be relatively straightforward. For example, if this is the current `test/runtests.jl` file:

```julia
using HelloWorld
using Test

@testset "HelloWorld.jl" begin
    @testset "foo" begin
        @test HelloWorld.foo(1) == 2
    end
    @testset "bar" begin
        @test HelloWorld.bar(2) == 6
    end
    @testset "baz" begin
        @test HelloWorld.baz(3) == 12
    end
end
```

Then you could modify it to look like this:

```julia
using HelloWorld
using Test

@testset "HelloWorld.jl" begin
    if isempty(ARGS) || "all" in ARGS
        all_tests = true
    else
        all_tests = false
    end
    if all_tests || "foo" in ARGS
        @testset "foo" begin
            @test HelloWorld.foo(1) == 2
        end
    end
    if all_tests || "bar" in ARGS
        @testset "bar" begin
            @test HelloWorld.bar(2) == 6
        end
    end
    if all_tests || "baz" in ARGS
        @testset "baz" begin
            @test HelloWorld.baz(3) == 12
        end
    end
end
```

And now if you run `Pkg.test("HelloWorld"; test_args = ["foo", "baz"])`, then only the `foo` and `baz` testsets will be run.

KristofferC commented:
I like it. We'll have to discuss a bit on the exact API for the command line args but this feels pretty ok.

bors try

KristofferC commented:
bors try

KristofferC commented:
We probably also want to be able to give julia startup command line options. For example `--track-allocations`.

fredrikekre commented on src/Operations.jl:
```diff
@@ -1228,7 +1229,10 @@ end
 testfile(source_path::String) = joinpath(testdir(source_path), "runtests.jl")
-function test(ctx::Context, pkgs::Vector{PackageSpec}; coverage=false, test_fn=nothing)
+function test(ctx::Context, pkgs::Vector{PackageSpec}; coverage=false, test_fn=nothing, args=String[])
+    if isa(args, AbstractString)
```
I know this is how `Base.runtests` does this, but its ugly and I think we should restrict to `Vector{String}`.

DilumAluthge replied:
Makes sense to me. But could we do `AbstractVector{<:AbstractString}` instead of `Vector{String}`?

fredrikekre replied:
Sure.

DilumAluthge replied:
Done!

DilumAluthge replied:
(Will push shortly)

DilumAluthge commented:
> We probably also want to be able to give julia startup command line options. For example --track-allocations. So perhaps there should be julia_args, test_args or something?

I like `julia_args`

DilumAluthge commented:
I renamed `args` to `test_args`, and I added a new keyword argument `julia_flags`.

fredrikekre commented:
Is it not better with `julia_args` and `test_args`?
We should also think of a Pkg REPL syntax for this. I am thinking only `test_args` needs to be supported, since `julia_args` should not be so commonly used. Maybe something like
```
pkg> test Example -- arg1 arg2
```
or something, i.e. using `--` as a separator. This could be implemented in a separate PR of course.

DilumAluthge commented:
I renamed `julia_flags` to `julia_args`!

DilumAluthge commented:
> We should also think of a Pkg REPL syntax for this. I am thinking only test_args needs to be supported, since julia_args should not be so commonly used.

Could we support both? Maybe something like:
```julia
pkg> test Example --julia-args --track-allocation=all --test-args arg1 arg2
```

But yeah I definitely agree that we should implement this in a separate pull request.

StefanKarpinski commented:
That looks quite a bit worse to me. At that point, shouldn't you just be calling the test script manually?

fredrikekre commented:
I agree, we have not even had a feature request for customizing julia flags

DilumAluthge commented:
> That looks quite a bit worse to me. At that point, shouldn't you just be calling the test script manually?

Fair enough. So if we add REPL support, it will only be for test args.

fredrikekre commented on src/Operations.jl:
```diff
@@ -1130,10 +1130,14 @@ function free(ctx::Context, pkgs::Vector{PackageSpec})
-function gen_test_code(testfile::String; coverage=false)
+function gen_test_code(testfile::String;
+        coverage=false,
+        julia_args::AbstractVector{<:AbstractString}=String[],
```
Maybe this should be `AbstractVector{Cmd}`? And you specify it as
```
Pkg.test("Example"; julia_args = [`--startup-file=no`, `-O3`])
```
?

DilumAluthge replied:
I like it.

DilumAluthge replied:
Done.

fredrikekre commented on src/Operations.jl:
```diff
@@ -1130,10 +1130,14 @@ function free(ctx::Context, pkgs::Vector{PackageSpec})
     code = """
         $(Base.load_path_setup_code(false))
         cd($(repr(dirname(testfile))))
+        append!(empty!(ARGS), $(repr(test_args)))
```
Could `repr` here fail for non-`Vector{String}`? Maybe we should just collect the arguments to `Vector{String}` at the entry point.

DilumAluthge replied:
I've implemented your `julia_args = String[a for a in julia_args]`  and `test_args = String[a for a in test_args]` ideas from below, which should solve this.

DilumAluthge replied:
Done.

fredrikekre commented on src/Operations.jl:
```diff
@@ -1145,7 +1149,7 @@ function gen_test_code(testfile::String; coverage=false)
         --startup-file=$(Base.JLOptions().startupfile == 1 ? "yes" : "no")
         --track-allocation=$(("none", "user", "all")[Base.JLOptions().malloc_log + 1])
-        --eval $code
+        $(julia_args) --eval $code
```
```suggestion
        $(julia_args) 
        --eval $code
```

fredrikekre commented on src/API.jl:
```diff
@@ -229,7 +229,9 @@ test(pkg::Union{AbstractString, PackageSpec}; kwargs...) = test([pkg]; kwargs...
-              coverage=false, test_fn=nothing, kwargs...)
+              coverage=false, test_fn=nothing,
+              julia_args::AbstractVector{<:AbstractString}=String[],
+              test_args::AbstractVector{<:AbstractString}=String[], kwargs...)
```
I suggest to do something like
```
julia_args = String[a for a in julia_args]
test_args = String[a for a in test_args]
```
here at the entry point and use `Vector{String}` only for the internal methods. Or
```
julia_args = Cmd[a for a in julia_args]
```
if you go with the `Cmd` approach.

DilumAluthge replied:
Done.

fredrikekre commented on src/Operations.jl:
```diff
@@ -1130,10 +1130,14 @@ function free(ctx::Context, pkgs::Vector{PackageSpec})
-function gen_test_code(testfile::String; coverage=false)
+function gen_test_code(testfile::String;
+        coverage=false,
+        julia_args::AbstractVector{<:AbstractString}=String[],
```
```suggestion
        julia_args::Vector{String}=String[],
```
or
```suggestion
        julia_args::Vector{Cmd}=Cmd[],
```

fredrikekre commented on src/Operations.jl:
```diff
@@ -1228,7 +1232,10 @@ end
-function test(ctx::Context, pkgs::Vector{PackageSpec}; coverage=false, test_fn=nothing)
+function test(ctx::Context, pkgs::Vector{PackageSpec};
+        coverage=false, test_fn=nothing,
+        julia_args::AbstractVector{<:AbstractString}=String[],
```
```suggestion
        julia_args::Vector{String}=String[],
```
or
```suggestion
        julia_args::Vector{Cmd}=Cmd[],
```

fredrikekre commented on src/Operations.jl:
```diff
@@ -1228,7 +1232,10 @@ end
+function test(ctx::Context, pkgs::Vector{PackageSpec};
+        coverage=false, test_fn=nothing,
+        julia_args::AbstractVector{<:AbstractString}=String[],
+        test_args::AbstractVector{<:AbstractString}=String[])
```
```suggestion
        test_args::Vector{String}=String[])
```

fredrikekre commented on src/Pkg.jl:
```diff
@@ -139,13 +139,16 @@ const update = API.up
-    Pkg.test(; coverage::Bool=false)
-    Pkg.test(pkg::Union{String, Vector{String}; coverage::Bool=false)
-    Pkg.test(pkgs::Union{PackageSpec, Vector{PackageSpec}}; coverage::Bool=false)
+    Pkg.test(; coverage::Bool=false, julia_args::AbstractVector{<:AbstractString}=String[], test_args::AbstractVector{<:AbstractString}=String[])
```
This looks a bit ugly, maybe rewrite it as 
```
    Pkg.test(; kwargs...)
```
etc and explain allowed kwargs in a bulleted list below

DilumAluthge replied:
Done.

fredrikekre commented on test/sandbox.jl:
```diff
@@ -17,6 +17,27 @@ test_test(fn; kwargs...)       = Pkg.test(;test_fn=fn, kwargs...)
+        test_test(; test_args=["a", "b", "c"], julia_args=["--handle-signals=yes", "--warn-overwrite=no", "--math-mode=ieee"]) do
+            @test true
+        end
+    end end
```
Tests here should test the actual API; e.g. with `Pkg.test()`

DilumAluthge replied:
Done.

fredrikekre commented on test/test_packages/RuntestsArgs/Manifest.toml:
```diff
@@ -0,0 +1,2 @@
+# This file is machine-generated - editing it directly is not advised
```
rm

DilumAluthge replied:
Done.

fredrikekre commented on test/test_packages/RuntestsArgs/test/Manifest.toml:
```diff
@@ -0,0 +1,33 @@
+# This file is machine-generated - editing it directly is not advised
```
rm this file too?

DilumAluthge replied:
Done.

fredrikekre commented on test/test_packages/RuntestsArgs/test/runtests.jl:
```diff
@@ -0,0 +1,6 @@
+
+@testset "RuntestsArgs.jl" begin
+    @test all(ARGS .== RuntestsArgs.expected_arguments())
+end
```
I would also add tests for julia_args, e.g. by inspecting `Base.JLOptions()` for some values that you set from `Pkg.test(; julia_args = ...)`

DilumAluthge replied:
Done.

DilumAluthge commented:
I've made the requested changes!

Also, in the interest of "eating our own dog food", I've updated the `Pkg` test suite to support running specific testsets. I've [added an example](https://github.com/DilumAluthge/Pkg.jl/blob/db30f565e5c5737287f2aa17b7544fb0d794123f/.travis.yml#L36) to `.travis.yml` to demonstrate.

cmcaine commented:
> That looks quite a bit worse to me. At that point, shouldn't you just be calling the test script manually?

If you want to run a test script manually with julia I think you have to Pkg.add() all of the deps and extras for the package before it will work, so it could be a moderate pain to do.

Kind of off-topic:

Could solve that in another PR by exposing (if they are not already exposed) some functions for getting the right LOAD_PATH from a Project.toml (like activate, but including extras, and possibly not actually changing LOAD_PATH for you, just returning what you should use).

An extra use case for this is running ancillary test scripts (e.g. `test/compare_examples.jl` in Gadfly (it diffs binary outputs that have been saved to disk from master and the feature branch for regression tests)).

fredrikekre commented:
> Kind of off-topic:
> Could solve that in another PR by exposing (if they are not already exposed) some functions for getting the right LOAD_PATH from a Project.toml (like activate, but including extras, and possibly not actually changing LOAD_PATH for you, just returning what you should use).

This is exposed(ish) in Julia 1.2; you can there run tests as
```
julia --project=test test/runtests.jl
```

cmcaine commented:
As mentioned on slack, I think that would require copying the Project.toml to the test directory and making some manual changes first.

DilumAluthge commented:
I think what @fredrikekre is referring to is that in Julia 1.2 and higher, you specify the test dependencies in `test/Project.toml`, not in the main `Project.toml` file.

fredrikekre commented:
Looks pretty good now, but

> Also, in the interest of "eating our own dog food", I've updated the Pkg test suite to support running specific testsets. I've added an example to .travis.yml to demonstrate.

can we undo that? I don't think that is necessary and just results in a huge diff.

DilumAluthge commented:
> > Also, in the interest of "eating our own dog food", I've updated the Pkg test suite to support running specific testsets. I've added an example to .travis.yml to demonstrate.
> 
> can we undo that? I don't think that is necessary and just results in a huge diff.

Alright I think I've reverted all of those changes!

fredrikekre commented:
bors try

fredrikekre commented:
I simplified the tests, and fixed running the old code-path.

bors try

fredrikekre commented:
bors r+

rfourquet commented:
This is really neat, thanks! Also, my +1 to Pkg REPL support for `test_args`.
